### PR TITLE
skip compact when rev not change at period compact mode

### PR DESCRIPTION
**Pull request by qsyqian**
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

### CODE REVIEW

File: server/etcdserver/api/v3compactor/periodic.go [hunk 1]
```diff
@@ -100,7 +100,7 @@ func (pc *Periodic) Run() {
 	compactInterval := pc.getCompactInterval()
 	retryInterval := pc.getRetryInterval()
 	retentions := pc.getRetentions()
-
+	prev := int64(0)
```
Comment (serathius) [inline server/etcdserver/api/v3compactor/periodic.go:103]: This variable should be within the goroutine.

File: server/etcdserver/api/v3compactor/periodic.go [hunk 2]
```diff
@@ -140,6 +143,7 @@ func (pc *Periodic) Run() {
 			startTime := pc.clock.Now()
 			_, err := pc.c.Compact(pc.ctx, &pb.CompactionRequest{Revision: rev})
 			if err == nil || err == mvcc.ErrCompacted {
+				prev = rev
```
Comment (serathius) [inline server/etcdserver/api/v3compactor/periodic.go:146]: Please move it this to line 152, next to setting of the `lastSuccess`

**Review by serathius — CHANGES_REQUESTED** [review]
I think the change makes sense, however we should add a test.

File: server/etcdserver/api/v3compactor/periodic.go (hunk 1 above)
Comment (qsyqian) [reply server/etcdserver/api/v3compactor/periodic.go]: done

**Comment by qsyqian** [comment]
@serathius add ut, ptal, thanks

File: server/etcdserver/api/v3compactor/periodic.go [hunk 3]
```diff
@@ -121,16 +122,18 @@ func (pc *Periodic) Run() {
 					continue
 				}
 			}
-
-			if pc.clock.Now().Sub(lastSuccess) < baseInterval {
+			rev := pc.revs[0]
+			if rev == lastRevision {
+				continue
+			}
```
Comment (serathius) [inline server/etcdserver/api/v3compactor/periodic.go:128]:
```suggestion
```
Covered by line below

File: server/etcdserver/api/v3compactor/periodic_test.go [hunk 4]
```diff
@@ -172,3 +173,99 @@ func TestPeriodicPause(t *testing.T) {
 		t.Errorf("compact request = %v, want %v", a[0].Params[0], wreq.Revision)
 	}
 }
+
+func TestPeriodicSkipRevNotChangeHourly(t *testing.T) {
```
Comment (serathius) [inline server/etcdserver/api/v3compactor/periodic_test.go:177]: Why do we add a test for both hours and minutes? Is there a expected difference? From comments compaction didn't happens.

File: server/etcdserver/api/v3compactor/periodic_test.go [hunk 5]
```diff
@@ -172,3 +173,52 @@ func TestPeriodicPause(t *testing.T) {
...
+
+	// compaction doesn't happens at every interval since revision not change
+	for i := 0; i < 5; i++ {
+		for j := 0; j < intervalsPerPeriod; j++ {
+			rg.SetRev(int64(100))
+			rg.Wait(1)
+			fc.Advance(tb.getRetryInterval())
+		}
+
+		_, err := compactable.Wait(1)
```
Comment (serathius) [inline server/etcdserver/api/v3compactor/periodic_test.go:219]: Should we check if any new compaction requests were recorded?

Comment (qsyqian) [reply server/etcdserver/api/v3compactor/periodic_test.go:219]: yes, i added it.

**Review by serathius — APPROVED** [review]
cc @ptabor @spzala @ahrtr

**Review by ahrtr — APPROVED** [review]
LGTM

But it seems that all the test cases in period_test.go should can be refactored using subtests, because the workflow is similar and they can share majority of the source code.  We can address it in a  separate PR.

**Review by spzala — APPROVED** [review]
lgtm. Also, manually tested changes.
Thanks @qsyqian

**Pull request merged**